### DISCUSSION TITLE
How to inject custom attributes?

### DISCUSSION THREAD

**Discussion by ds5678** [mono/CppSharp#1804] [Q&A] [answered]
I can't figure out how to inject a custom attribute onto a type or member. In my case, I want to apply `SupportedOSPlatform` to a few Windows-specific functions and types.

**Comment by tritao** [comment]
I don't think this is supported out of the box atm, but should be a simple patch to add, PRs appreciated :+1:

**Reply by ds5678** [reply]
Can you give a simple explanation of what I would need to change to add support for this? I am not yet familiar with the code base, so any hints would be appreciated.

**Reply by tritao** [reply, marked as answer]
You might be in luck, looks like we have some kind of support for writing attributes per declaration:

Check these and let me know if it helps:

https://github.com/mono/CppSharp/blob/main/src/AST/Declaration.cs#L312
https://github.com/mono/CppSharp/blob/main/src/Generator/Generators/CSharp/CSharpSources.cs#L359

**Reply by ds5678** [reply]
Thank you so much. It worked great!

```cs
public void Postprocess(Driver driver, ASTContext ctx)
{
    foreach (var function in ctx.TranslationUnits.SelectMany(t => t.Functions))
    {
        function.Attributes.Add(new Attribute() { Type = typeof(SupportedOSPlatformAttribute), Value = "\"windows\"" });
    }
}
```
